File: hao/c++/1371-find-the-longest-substring-containing-vowels-in-even-counts.cpp
class Solution {
public:
    typedef struct {
        int min = -1;
        int max = -1;
    } Counter;
    int findTheLongestSubstring(string s) {
        uint8_t key = 0;
        Counter counter[32];
        int length = s.length();
        int i,j;
        for (i = 0; i < length; i++) {
            switch(s[i]) {
                case 'a': {
                    key ^= 1;
                }
                case 'e': {
                    key ^= 2;
                }
                case 'i': {
                    key ^= 4;
                }
                case 'o': {
                    key ^= 8;
                }
                case 'u': {
                    key ^= 16;
                }
                default:
                    break;
            }
            counter[key].max = i;
            if (counter[key].min == -1) {
                counter[key].min = i;
            }
        }
        length = -1;
        counter[0].min = -1;
        for (i = 0; i < 32; i++) {
            j = counter[i].max - counter[i].min;
            if (j > length) {
                length = j;
            }
        }
        return length;
    }
};

File: tao/python/1371-find-the-longest-substring-containing-vowels-in-even-counts.py
class Solution:
    def findTheLongestSubstring(self, s: str) -> int:
        l = list(s)
        vowels = {'a': 1, 'e': 2,'i': 3,'o': 4,'u': 5}
        
        record = ans = 0
        hash_map = {record:-1}

        for i, c in enumerate(l):
            if c in vowels: 
                record ^= (1<< vowels[c])
            
            if record==0: ans = max(ans, i+1)
            else:   
                if record not in hash_map:
                    hash_map[record] = i
                else:
                    ans = max(ans, i-hash_map[record]) 
        return ans
                